share/wake/lib/system/job.wake

```diff
@@ -440,108 +736,189 @@ export def defaultRunner: Runner = fuseRunner
-              Pair (Pass proxy) inFile
-  def post = match _
-    Pair (Fail f) _ = Fail f
-    Pair (Pass (RunnerOutput _ _ (Usage x _ _ _ _ _))) inFile if x != 0 =
-      Fail (makeError "Non-zero exit status ({str x}) for JSON runner {script} on {inFile}")
-    Pair (Pass _) inFile =
-      def unlink f =
-        def fn _ = prim "unlink"
-        match (getenv "WAKEBOX_KEEP_IO_FILES")
-          Some "1" = Unit
-          _        = fn f
-      def outFile = replace `\.in\.json$` ".out.json" inFile
-      def json = parseJSONFile (Path outFile)
-      def _ = unlink inFile
-      match json
-        Fail f = Fail f
-        Pass content =
-          def _ = unlink outFile
-          def field name = match _ _
-             _ (Fail f) = Fail f
-             None (Pass fn) = Fail "{script} produced {outFile}, which is missing usage/{name}"
-             (Some x) (Pass fn) = Pass (fn x)
-          def usage = content // `usage`
-          def usageResult =
-            Pass (Usage _ _ _ _ _ _)
-            | field "status"   (usage // `status`   | getJInteger)
-            | field "runtime"  (usage // `runtime`  | getJDouble)
-            | field "cputime"  (usage // `cputime`  | getJDouble)
-            | field "membytes" (usage // `membytes` | getJInteger)
-            | field "inbytes"  (usage // `inbytes`  | getJInteger)
-            | field "outbytes" (usage // `outbytes` | getJInteger)
-          def getK exp = content // exp | getJArray | getOrElse Nil | mapPartial getJString
-          match usageResult
-            Fail f = Fail (makeError f)
-            Pass usage = Pass (RunnerOutput (getK `inputs`) (getK `outputs`) usage)
-  makeRunner "json-{script}" score pre post localRunner
+    def rawScript =
+        plan.getJSONRunnerPlanRawScript
+    def extraArgs =
+        plan.getJSONRunnerPlanExtraArgs
+    def extraEnv =
+        plan.getJSONRunnerPlanExtraEnv
+    def score =
+        plan.getJSONRunnerPlanScore
+    def estimate =
+        plan.getJSONRunnerPlanEstimate
+
+    def script =
+        which (simplify rawScript)
+    def ok =
+        access script xOK
+    def pre =
+        match _
+            Fail f =
+                Pair (Fail f) ""
+            _ if ! ok =
+                Pair (Fail (makeError "Runner {script} is not executable")) ""
+            Pass (RunnerInput label command visible environment directory stdin res prefix record) =
+                def Usage status runtime cputime membytes inbytes outbytes =
+                    record
+                def pmkdir m p =
+                    prim "mkdir"
+                def pwrite m p d =
+                    prim "write"
+                def json =
+                    JObject
+                    (
+                        "label" :-> JString label,
+                        "command"
+                        :-> command
+                        | map JString
+                        | JArray,
```

Comment:
I think the human perspective categorizes symbolic operators into distinct categories, which the AST doesn't necessarily capture.

Reply:
What's super confusing here is that the operator precedence is not respected.
The expression here is `"command" :-> (command | map JSrting | JArray)` and the formatting totally hides that.